tamr_client/datasets/dataframe.py

```diff
@@ -11,28 +12,52 @@
 
 
 def upsert(
-    session: tc.Session, dataset: tc.Dataset, df: pd.DataFrame, *, primary_key_name: str
+    session: tc.Session,
+    dataset: tc.Dataset,
+    df: pd.DataFrame,
+    *,
+    primary_key_name: Optional[str] = None,
 ) -> JsonDict:
     """Upserts a record for each row of `df` with attributes for each column in `df`.
     Args:
         dataset: Dataset to receive record updates
         df: The DataFrame containing records to be upserted
-        primary_key_name: The primary key of the dataset.  Must be a column of `df`.
+        primary_key_name: The primary key of the dataset.  Must be a column of `df`. By default the key_attribute_name of dataset
 
     Returns:
-            JSON response body from the server
+        JSON response body from the server
 
     Raises:
         requests.HTTPError: If an HTTP error is encountered
-        PrimaryKeyNotFound: If `primary_key_name` is not a column in `df`.
+        PrimaryKeyNotFound: If `primary_key_name` is not a column in `df` or the index of `df`
+        ValueError: If `primary_key_name` matches both a column in `df` and the index of `df`
     """
-    if primary_key_name not in df.columns:
-        raise tc.PrimaryKeyNotFound(
-            f"Primary key: {primary_key_name} is not in DataFrame column names: {df.columns}"
+    if not primary_key_name:
+        primary_key_name = dataset.key_attribute_names[0]
+
+    if primary_key_name in df.columns and primary_key_name == df.index.name:
+        raise ValueError(
+            f"Index {primary_key_name} has the same name as column {primary_key_name}"
         )
+
+    if primary_key_name not in df.columns:
+        # if `df.index.name` is the primary key, create column from `df.index`
+        if primary_key_name == df.index.name:
+            df.insert(0, df.index.name, df.index)
```

Comment:
We should not mutate the passed in dataframe. Instead, we should make a copy and perform mutations on that.

---

Relatedly, the approach here seems to be:
- For datasets with the primary key as a column, do nothing.
- For datasets with the primary key as the index, make the index into a column.

I suggest inverting this approach to:
- For datasets with the primary key as a column, make the column into the index
- For datasets with the primary key as the index, do nothing.

Making the column into an index should be done on a *copy* of the dataframe to avoid mutations to the passed in dataframe.

Then we will have the correct semantics plus we can use *both* values of each item yielded by `iterrows`:

```python
serialized_records = (
    {**row.to_json(), primary_key_name: pk}
    for pk, row in df.iterrows()
) 
```

Reply:
Note: making column into index (and doing this without mutation the original dataframe):

```python
index_df = df.set_index(primary_key_name)
```

Note that `drop` param defaults to `True` and `inplace` defaults to `False` which both work in our favor here.

Reply:
That sounds good to me.  I needed to adjust the line setting `serialized_records` a bit, but this seems to work well.